rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+       at test.nix:1:1:
+
+            1| https://nixos.org
+             | ^
+
+"https://nixos.org"
+
+warning: The following deprecated features were used:
+  - url-literal (httsp://nixos.org), 1 time
+
+  Add `--lang-warn-verbose` to show all occurrences
+  Use `--lang-no-warn=url-literal` to disable this warning.
+  Use `--lang-error=url-literal` to issue errors instead of warnings.
+```
+
+# Prior art
+
+- [Perl `use VERSION`]
+
+  [Perl `use VERSION`]: https://perldoc.perl.org/functions/use#use-VERSION
+
+  Many similarities, with versions declared per file and having to deal with interoperability.
+
+- [Rust `edition` field]
+
+  Rust has an easier problem to solve.
+  Cargo files are written in TOML, so the `edition` information does not have to be part of Rust itself.
+
+  [Rust `edition` field]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field
+
+- [Haskell language extensions]
+
+  Haskell allows enabling separate language features per file.
+
+  [Haskell language extensions]: https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/intro.html
+
+- JaveScript modules
+    - .cjs and .mjs extensions for commonjs/es-modules syntax variants
+    - `function() { "use strict"; return 10 }`
+
+- [Go language changes proposal](https://github.com/golang/proposal/blob/master/design/28221-go2-transitions.md#language-changes)
+
+  The document features a discussion of multiple approaches to evolve the language without breaking existing code, and comes to the similar conclusions as our proposal, mainly:
+  - One needs a way to specify the language version
+    - Interestingly, Go developers decided to equate the language version with the compiler release version, despite admitting that this may be confusing
+  - Keep support for code written in older language versions
+  - Breaking down into features that can be enabled separately is not practical
+
+- [Flakes `edition` field]
+
+  There had been an attempt to include an `edition` field into the Flakes schema.
+  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.
+
+  [Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6
+
+# Future work
+[future]: #future-work
+
+- Define rules deciding when a change to the language is appropriate, to avoid version proliferation and limit complexity of implementations.
+- Define a roadmap to introduce the next language versions, for example:
+  - Version 7 commits to changes that do not require additional work on Nixpkgs:
+    - Introduce the version declaration, required to distinguish versions 6 and 7
+    - Deprecate URL literals
+    - Deprecate the `let-body` syntax
+    - Drop support for leading zeroes on integers
+    - Formalise change in float representation
+  - Version 8 is released at least a year after version 7, and only when version annotations according to version 7 are fully supported in Nixpkgs
```

Comment:
Putting a timeline on version 8 makes me doubt the statement that “should remain a rare exception” reflects the authors intentions.

Reply:
This isn't a timeline but a lower bound, used in the example to illustrate the scale. I wouldn't know how such changes could even be done that fast. We can remove that part, no problem.

```suggestion
  - Version 8 is released only when version annotations according to version 7 are fully supported in Nixpkgs
```